### PR TITLE
added .status.observedGeneration to camel K CRDs

### PR DESCRIPTION
**Pull request by SubhasmitaSw**
fixes: issue #3182 

cc @astefanutti

### CODE REVIEW

**Review by tadayosi — APPROVED** [review]

**Review by squakez — COMMENTED** [review]
Nice, thanks. However I think we're still missing the update part. We should check in the reconciliation cycle and update accordingly.

**Comment by squakez** [comment]
We should also provide some basic E2E test if possible.

**Comment by SubhasmitaSw** [comment]
> Nice, thanks. However I think we're still missing the update part. We should check in the reconciliation cycle and update accordingly.

For this part, @astefanutti was about to share some background references to me regarding `metadata.generation` field, however he seems to be occupied so I'll be glad if you could provide some information on proceeding with this.

**Comment by squakez** [comment]
I am not very familiar with this part, but I think you can have a look at the update function, which should be in charge to set that value: https://github.com/apache/camel-k/blob/main/pkg/controller/build/build_controller.go#L74

File: pkg/controller/kamelet/kamelet_controller.go [hunk 1]
```diff
@@ -200,6 +200,9 @@ func (r *reconcileKamelet) Reconcile(ctx context.Context, request reconcile.Requ
 		break
 	}
 
+	// Update the metadata.generation to avoid conflicts with other reconcile operations
+	target.Status.ObservedGeneration = instance.GetGeneration()
```
Comment (astefanutti) [inline pkg/controller/kamelet/kamelet_controller.go:204]: This must be added just before:

```
if err := r.client.Status().Patch(ctx, target, ctrl.MergeFrom(&instance)); err != nil {
```

Line 181.

File: pkg/controller/kameletbinding/kamelet_binding_controller.go [hunk 2]
```diff
@@ -214,6 +214,8 @@ func (r *ReconcileKameletBinding) Reconcile(ctx context.Context, request reconci
 			break
 		}
 	}
+	// Update the metadata.Generation to avoid conflicts with other clients updating the resource at the same time we are reconciling it.
+	target.Status.ObservedGeneration = instance.Generation
```
Comment (astefanutti) [inline pkg/controller/kameletbinding/kamelet_binding_controller.go:218]: This must be added just before line 195.

File: pkg/controller/integrationplatform/integrationplatform_controller.go [hunk 3]
```diff
@@ -173,6 +173,9 @@ func (r *reconcileIntegrationPlatform) Reconcile(ctx context.Context, request re
 	target := instance.DeepCopy()
 	targetLog := rlog.ForIntegrationPlatform(target)
 
+	// Update the metadata.generation to avoid conflicts with other reconcile operations
+	target.Status.ObservedGeneration = instance.Generation
```
Comment (astefanutti) [inline pkg/controller/integrationplatform/integrationplatform_controller.go:177]: I'd recommend to add it just before line 195.

File: pkg/controller/integrationkit/integrationkit_controller.go [hunk 4]
```diff
@@ -303,6 +303,7 @@ func (r *reconcileIntegrationKit) Reconcile(ctx context.Context, request reconci
 
 func (r *reconcileIntegrationKit) update(ctx context.Context, base *v1.IntegrationKit, target *v1.IntegrationKit) (reconcile.Result, error) {
 	dgst, err := digest.ComputeForIntegrationKit(target)
+	target.Status.ObservedGeneration = base.Generation
```
Comment (astefanutti) [inline pkg/controller/integrationkit/integrationkit_controller.go:306]: This is better moved after line 311.

File: pkg/controller/integration/integration_controller.go [hunk 5]
```diff
@@ -338,6 +338,7 @@ func (r *reconcileIntegration) Reconcile(ctx context.Context, request reconcile.
 }
 
 func (r *reconcileIntegration) update(ctx context.Context, base *v1.Integration, target *v1.Integration) (reconcile.Result, error) {
+	target.Status.ObservedGeneration = base.Generation
```
Comment (astefanutti) [inline pkg/controller/integration/integration_controller.go:341]: I'd suggest to move it after line 347.

File: pkg/apis/camel/v1alpha1/kamelet_binding_types.go [hunk 6]
```diff
@@ -94,6 +94,8 @@ type EndpointProperties struct {
 
 // KameletBindingStatus specify the status of a binding
 type KameletBindingStatus struct {
+	// ObservedGeneration is the generation observed by the KameletBinding controller.
```
Comment (astefanutti) [inline pkg/apis/camel/v1alpha1/kamelet_binding_types.go:97]:
```suggestion
	// ObservedGeneration is the most recent generation observed for this KameletBinding.
```

File: pkg/apis/camel/v1/integrationplatform_types.go [hunk 7]
```diff
@@ -52,7 +52,8 @@ type IntegrationPlatformResourcesSpec struct {
 // IntegrationPlatformStatus defines the observed state of IntegrationPlatform
 type IntegrationPlatformStatus struct {
 	IntegrationPlatformSpec `json:",inline"`
-
+	// ObservedGeneration is the most recent generation observed by the controller. It corresponds to the controller's `metadata.generation` and is updated on mutation by the API Server.
```
Comment (astefanutti) [inline pkg/apis/camel/v1/integrationplatform_types.go:55]:
```suggestion
	// ObservedGeneration is the most recent generation observed for this IntegrationPlatform.
```

File: pkg/apis/camel/v1/integrationkit_types.go [hunk 8]
```diff
@@ -67,6 +67,8 @@ type IntegrationKitSpec struct {
 
 // IntegrationKitStatus defines the observed state of IntegrationKit
 type IntegrationKitStatus struct {
+	// ObservedGeneration is the last generation change we observed. This is used to determine if we need to refresh the status of the integration kit.
```
Comment (astefanutti) [inline pkg/apis/camel/v1/integrationkit_types.go:70]:
```suggestion
	// ObservedGeneration is the most recent generation observed for this IntegrationKit.
```

File: pkg/apis/camel/v1/integration_types.go [hunk 9]
```diff
@@ -85,6 +85,9 @@ type IntegrationSpec struct {
 
 // IntegrationStatus defines the observed state of Integration
 type IntegrationStatus struct {
+	// ObservedGeneration is the most recent generation observed for this Integration.
```
Comment (astefanutti) [inline pkg/apis/camel/v1/integration_types.go:88]:
```suggestion
	// ObservedGeneration is the most recent generation observed for this Integration.
```

File: pkg/apis/camel/v1/build_types.go [hunk 10]
```diff
@@ -146,6 +146,8 @@ type S2iTask struct {
 
 // BuildStatus defines the observed state of Build
 type BuildStatus struct {
+	//ObservedGeneration is the most recent generation observed for this Build.
```
Comment (astefanutti) [inline pkg/apis/camel/v1/build_types.go:149]:
```suggestion
	// ObservedGeneration is the most recent generation observed for this Build.
```

**Review by astefanutti — CHANGES_REQUESTED** [review]

File: pkg/controller/kamelet/kamelet_controller.go [hunk 11]
```diff
@@ -178,6 +178,9 @@ func (r *reconcileKamelet) Reconcile(ctx context.Context, request reconcile.Requ
 		}
 
 		if target != nil {
+			// Update the metadata.generation to avoid conflicts with other reconcile operations.
```
Comment (astefanutti) [inline pkg/controller/kamelet/kamelet_controller.go:181]: That comment can be removed.

File: pkg/controller/kameletbinding/kamelet_binding_controller.go [hunk 12]
```diff
@@ -192,6 +192,9 @@ func (r *ReconcileKameletBinding) Reconcile(ctx context.Context, request reconci
 			}
 
 			if target != nil {
+				// Update the metadata.Generation to avoid conflicts with other clients updating the resource at the same time we are reconciling it.
```
Comment (astefanutti) [inline pkg/controller/kameletbinding/kamelet_binding_controller.go:195]: That comment can be removed.

File: pkg/controller/integrationplatform/integrationplatform_controller.go [hunk 13]
```diff
@@ -189,6 +189,9 @@ func (r *reconcileIntegrationPlatform) Reconcile(ctx context.Context, request re
 			}
 
 			if target != nil {
+				// Update the metadata.generation to avoid conflicts with other reconcile operations
```
Comment (astefanutti) [inline pkg/controller/integrationplatform/integrationplatform_controller.go:192]: That comment can be removed.

**Review by astefanutti — COMMENTED** [review]
Added a couple of comments, also please regenerate the CRDs and we should be good to go :)

File: pkg/apis/camel/v1/integration_types.go [hunk 14]
```diff
@@ -85,6 +85,9 @@ type IntegrationSpec struct {
...
+	// It corresponds to the Integration's generation, which is updated on mutation by the API Server.
```
Comment (astefanutti) [inline pkg/apis/camel/v1/integration_types.go:89]: That line can be removed.

**Review by astefanutti — APPROVED** [review]
Thanks!

**Comment by astefanutti** [comment]
There are some YAKS tests failing, that requires a change in YAKS itself. I've created citrusframework/yaks#403. Once it's fixed, that PR can be merged.

**Comment by christophd** [comment]
YAKS has a pinned prerelease available [0.11.0-202204210858](https://github.com/citrusframework/yaks/releases/tag/0.11.0-202204210858) that should fix the issue.

@SubhasmitaSw to use the fixed YAKS version please update the version in the GitHub actions workflow to `0.11.0-202204210858`:
https://github.com/apache/camel-k/blob/main/.github/actions/e2e-knative-yaks/action.yml#L81

**Comment by astefanutti** [comment]
Let's merge this. The `TestKitTimerToLogFullNativeBuild` test failure is already present in main branch.

**Comment by astefanutti** [comment]
Thanks!

**Pull request merged**